### PR TITLE
Enables support for multiple plural forms

### LINKED ISSUE #6: Support for multiple plural forms (one, few, many, plural, ...)
Hi I'd like to work on it, I think we can follow [this pattern](https://www.i18next.com/translation-function/plurals#languages-with-multiple-plurals)

can I?

### PR DESCRIPTION
**Pull request by hjJunior**
fix #6 

## The definition of localizations

```json
{
  "clicked": "You clicked {{count}} time!",
  "clicked_0": "No clicks",
  "clicked_2": "Double tap",
  "clicked_plural": "You clicked {{count}} times!"
}
```

### CODE REVIEW

File: test/plural_resolver_test.dart [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    test('when count is equal 1 returns empty string', () {
+      const count = 1;
+      final result = subject.pluralize(
+        count,
+        I18NextOptions.base,
+        checkerKeyFnc,
+      );
+
+      verifyZeroInteractions(checkerKeyFnc);
+      expect(result, equals(''));
```
Comment (williamhjcho) [inline test/plural_resolver_test.dart:31]: Instead of the mock, you could use the `expectAsync1` instead.

File: test/plural_resolver_test.dart [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import 'package:mockito/mockito.dart';
+
+abstract class FakeCheckerKeyFunction {
+  bool call(String pluralModifier);
+}
+
+class MockCheckerKeyFunction extends Mock implements FakeCheckerKeyFunction {}
+
+void main() {
+  group('PluralResolver', () {
```
Comment (williamhjcho) [inline test/plural_resolver_test.dart:13]: This file is already for the plural resolver, remove this group

File: test/plural_resolver_test.dart [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    MockCheckerKeyFunction checkerKeyFnc;
+    PluralResolver subject;
+
+    setUpAll(() {
+      checkerKeyFnc = MockCheckerKeyFunction();
+      subject = PluralResolver();
+    });
+
```
Comment (williamhjcho) [inline test/plural_resolver_test.dart:21]: Missing test when count is null

File: test/plural_resolver_test.dart [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    test('when count is equal 1 returns empty string', () {
```
Comment (williamhjcho) [inline test/plural_resolver_test.dart:22]: We're using a more BDD style approach for the tests

Unless the expected results are ambigous, you can omit the result on the test name.

`"when X" -> expect(X, ...)`
`"given Y" -> expect(function(Y), ...)`

The expectations should be clear (this helps keeping the tests as minimal as possible)

File: test/plural_resolver_test.dart [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        count,
+        I18NextOptions.base,
+        checkerKeyFnc,
+      );
+
+      verifyZeroInteractions(checkerKeyFnc);
+      expect(result, equals(''));
+    });
+
+    group('when count is different than 1', () {
```
Comment (williamhjcho) [inline test/plural_resolver_test.dart:34]: Missing count zero, it is more to keep it explicit what a zero should return or not

File: lib/src/plural_resolver.dart [hunk 6]
```diff
@@ -2,25 +2,38 @@ import 'dart:ui';
 
 import 'options.dart';
 
+/// A callback function to determine if exists
+/// the key for an specifc plural modifier
+typedef CheckerKeyFunction = bool Function(
+  String pluralModifier,
+);
```
[the remark this answers is not in the document]
Comment (hjJunior) [reply lib/src/plural_resolver.dart:9]: `KeyExists` sounds good to me

File: test/plural_resolver_test.dart (hunk 3 above)
Comment (hjJunior) [reply test/plural_resolver_test.dart:21]: IMHO I don't think we need to test (specifically for PluralResolver) when the count is null since we have this validation on the method `translateKey`

```dart
    final needsPlural = count != null;

    String pluralSuffix;
    if (needsPlural) {
      pluralSuffix = pluralResolver.pluralize(
        count,
        options,
        (pluralModifier) => _hasPluralKey(
          pluralModifier,
          baseKey: key,
          locale: locale,
          namespace: namespace,
          options: options,
        ),
      );
    }
``` 

Maybe it should be good to add this requested test to `Translator` class file tests, sound good?

File: lib/src/plural_resolver.dart [hunk 7]
```diff
@@ -2,25 +2,38 @@ import 'dart:ui';
...
-    // TODO: add locale based rules
-    return -1;
+  String _pluralKey(
+    int count,
+    I18NextOptions options,
+    CheckerKeyFunction find,
+  ) {
+    final baseKey = '${options.pluralSeparator}${options.pluralSuffix}';
+    final specicPluralKey = '$baseKey${options.pluralSeparator}$count';
+    final existSpecicPlural = find(specicPluralKey);
```
[the remark this answers is not in the document]
Comment (hjJunior) [reply lib/src/plural_resolver.dart:35]: Oh sorry about the locale-based, I confess that I don't have too much information about that, and about the issue, I'll work to fix it and make on the same specification from the official i18next

File: test/plural_resolver_test.dart (hunk 1 above)
Comment (hjJunior) [reply test/plural_resolver_test.dart:31]: What do u think to leave this to an 'integration' tests?

File: lib/src/plural_resolver.dart [hunk 8]
```diff
@@ -2,25 +2,38 @@ import 'dart:ui';
...
+    int count,
+    I18NextOptions options,
+    KeyExists validateKey,
+  ) {
+    final shouldLookForPluralKeys = count != 1;
+    const resultForNonRequiredPlural = '';
+
+    return shouldLookForPluralKeys
+        ? _pluralKey(count, options, validateKey)
+        : resultForNonRequiredPlural;
```
Comment (williamhjcho) [inline lib/src/plural_resolver.dart:25]: Something that came to mind now, with these changes there's the `count == 1` issue, right?

`key_1` is not allowed, since it is ignored by the early exit here.

Any thoughts on how to avoid that? Or to in some way mitigate the issue?

File: test/plural_resolver_test.dart (hunk 3 above)
Comment (williamhjcho) [reply test/plural_resolver_test.dart:31]: Yes, you're right.

Since the plural resolver is not public, then the only caller (translator) already does the initial checks/assertions early on.

File: test/plural_resolver_test.dart (hunk 1 above)
Comment (williamhjcho) [reply test/plural_resolver_test.dart]: Integration tests don't have direct access to the function parameter.

The suggestion is only to remove the usage of a mocking class altogether:
So for this test, it would be:

```dart
...
final result = subject.pluralize(..., expectAsync1(null, count: 0));
expect(result, equals(''));
```

Comment (williamhjcho) [reply test/plural_resolver_test.dart]: I'm only suggesting it because of the repeated pattern of mocking a method with a single value then expecting if it was called or not (which is what expectAsync is for)

File: lib/src/plural_resolver.dart (hunk 8 above)
Comment (hjJunior) [reply lib/src/plural_resolver.dart:25]: :thinking: oh yeah, I missed this one, actually, I've decided this skipping a little about the i18next semantic

In my perspective, from one side, it should be good to keep the same semantic of i18next, it will be easy to learn how to use this package

Btw, on another side, it looks strange if the user defines this specific localization JSON

```json
{
  "clicked": "You clicked {{count}} time!",
  "clicked_1": "One click",
  "clicked_plural": "You clicked {{count}} times!"
}
```

Should we get the key with `_1` when count is equal 1?
What are ur thoughts?

Comment (hjJunior) [reply lib/src/plural_resolver.dart:25]: I changed the code allowing bigger priority for keys with suffix `_1`

```dart
    final isPlural = count != 1;
    final hasSpecificPlural = _hasSpecificPlural(count, options, validateKey);
    const resultForNonRequiredPlural = '';

    if (hasSpecificPlural) {
      return '${options.pluralSeparator}$count';
    }

    return isPlural ? '_plural' : resultForNonRequiredPlural;
``` 

If you agree with this approach I'll push the code for it

Comment (williamhjcho) [reply lib/src/plural_resolver.dart:25]: The usage of multiple-plurals on i18next is only allowed to some locales (e.g. arabic), while others use the simple binary plural rule (`_plural`).

Maybe we should fix the special rule to known locales after all? (only the ones that require it)

Comment (hjJunior) [reply lib/src/plural_resolver.dart:25]: Yep, I think is a good idea, I need to search a little more about it, do you have this list of locales?

Comment (williamhjcho) [reply lib/src/plural_resolver.dart:25]: no, but the official doc has it (or at least in their source code)

**Comment by lynn** [comment]
Hi there!

As far as I understand it, the change proposed by this PR goes against the meaning `i18next` assigns to e.g. `key_2`.

`key_2` does **not** mean: “use this localization when `count` equals `2`”.

Instead, `key_2` **always** means: “use this string when [the plural-resolver for this locale](https://github.com/i18next/i18next/blob/4bfa7a3ace9d5eb6f7dee2fe4640b918242ba441/src/PluralResolver.js#L41) decides on index `2` for the given count.”

The meaning of that index is locale-dependent: For example, in Russian, `key_2` means `count % 10 == 0 || count % 10 >= 5 || (count >= 10 && count <= 20)` and indicates that the translation should use the genitive plural. Importantly, `count == 2` maps to `key_1`, _not_ `key_2`. This contradicts the significance this PR wants to assign to “`key_2`”.

(Actually, `key` and `key_plural` are just “simplified” forms of `key_0` and `key_1`, that are used in certain locales.)

If you want to do something like what this PR achieves, the i18next docs suggest the [interval plural preprocessor](https://www.i18next.com/translation-function/plurals#interval-plurals).

**Comment by lynn** [comment]
I made a PR #13 that ports the behavior from i18next for all locales it supports.

**Comment by williamhjcho** [comment]
Superseeded by #13
